Remove ParseFieldMatcher from AbstractXContentParser

cbuescher opened the pull request:
Currently we are able to set a ParseFieldMatcher on XContentParsers, mainly to conveniently carry it around to be available where the
actual parsing happens. This was just recently introduced together with ObjectParser so that ObjectParser can make use of deprecation
logging and throwing errors while parsing.

This however is trappy because we create parsers in so many places in the code and it is easy to forget setting the right ParseFieldMatcher.
Instead we should hold the ParseFieldMatcher only in the parse contexts (e.g. QueryParseContext).

This PR removes the ParseFieldMatcher from XContentParser. ObjectParser can still make use of it because we can make the otherwise unbounded
`context` type to extend an interface that makes sure contexts used in ObjectParser can supply a ParseFieldMatcher. Contexts in ObjectParser
are now no longer optional, but it is sufficient to pass in a small lambda expression in places where no other context is available.

Relates to #17417

cbuescher committed:
Remove ParseFieldMatcher from AbstractXContentParser

Currently we are able to set a ParseFieldMatcher on XContentParsers,
mainly to conveniently carry it around to be available where the
actual parsing happens. This was just recently introduced together
with ObjectParser so that ObjectParser can make use of deprecation
logging and throwing errors while parsing.

This however is trappy because we create parsers in so many places in
the code and it is easy to forget setting the right ParseFieldMatcher.
Instead we should hold the ParseFieldMatcher only in the parse contexts
(e.g. QueryParseContext).

This PR removes the ParseFieldMatcher from XContentParser. ObjectParser
can still make use of it because we can make the otherwise unbounded
`context` type to extend an interface that makes sure contexts used in
ObjectParser can supply a ParseFieldMatcher. Contexts in ObjectParser
are now no longer optional, but it is sufficient to pass in a small
lambda expression in places where no other context is available.

Relates to #17417

nik9000 commented:
I genuinely preferred the old way but if it is too easy to break I can live with this.

javanna commented on core/src/main/java/org/elasticsearch/cluster/routing/AllocationId.java:
```diff
@@ -198,6 +201,6 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
     public static AllocationId fromXContent(XContentParser parser) throws IOException {
-        return ALLOCATION_ID_PARSER.parse(parser, new AllocationId.Builder()).build();
+        return ALLOCATION_ID_PARSER.parse(parser, new AllocationId.Builder(), () -> ParseFieldMatcher.STRICT).build();
```
I see how using strict is bw compatible as we had hardcoded STRICT in AbstractXContentParser. That said, that was wrong in the first place. We should make sure that we have some kind of parse context that is at least a ParseFieldMatcherSupplier every single time we parse anything. Callers of these static parse methods should provide it, and ideally these STRICT and EMPTY constants should be moved to test classes. Hopefully those callers will have settings available to understand whether we are in strict mode or not, or use the same defaults everywhere if settings are not available (not like we do now that we randomly use STRICT or EMPTY here and there). This can be a follow-up, it is a bigger change, but it seems easier and cleaner to achieve after this PR, which is great!

cbuescher replied:
I tracked this one back to the place where the parser that gets passed down is created. The call-stack here is something like:
MetaDataStateFormat#loadLatestState -> ShardStateMetaData#fromXContent -> Allocation#fromXContent. 

I think we use ObjectParser a lot now in situations there when we initialize the parsing process we don't really have the settings available that we need to decide if we want to parse strict or without deprecation checks. This observation is indepentend of whether we want to store the ParseFieldMatcher in the parser or some sort of context. 
In this case I think we should default to STRICT because these json objects seem to be only created internally, so we should control that we never output anything deprecated.

javanna replied:
I figured that as well. this seems something very internal that we only generate and parse, so I tend to agree. although having these cases make these constants needed besides tests, which means that they will be misused like they are now. we can discuss this again later when we try to address this problem more closely, not directly related to this PR.

javanna commented on core/src/main/java/org/elasticsearch/common/ParseFieldMatcherSupplier.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * @return the parseFieldMatcher
+     */
+    ParseFieldMatcher parseFieldMatcher();
```
nit: can we call it getParseFieldMatcher? now that we have an interface it should be easy to rename all the existing impls at the same time. If you feel like it should be a follow-up, I am fine with that.

cbuescher replied:
will try, if this touches too many additional files I would probably do that in a follow up.

javanna commented on core/src/test/java/org/elasticsearch/common/xcontent/ObjectParserTests.java:
```diff
@@ -34,6 +35,8 @@
 
 public class ObjectParserTests extends ESTestCase {
 
+    private static ParseFieldMatcherSupplier strictParsing = () -> ParseFieldMatcher.STRICT;
```
final as well?

javanna replied:
also if it's a constant it should probably be STRICT_PARSING instead? but why do we need this constant again? :)

cbuescher replied:
Will rename. The constant is only here to not having to repeat the lambda everywhere in tests in this class.

javanna commented on core/src/test/java/org/elasticsearch/search/sort/SortParserTests.java:
```diff
@@ -47,7 +47,7 @@ public void testGeoDistanceSortParserManyPointsNoException() throws Exception {
         sortBuilder.field("order", "desc");
         sortBuilder.field("unit", "km");
-        sortBuilder.field("sort_mode", "max");
+        sortBuilder.field("mode", "max");
```
oh these changes seem to be a side effect of setting STRICT somewhere? But I thought we had STRICT already in the base class parser. Which change is causing this then?

cbuescher replied:
This is an example of how things are broken right now, the parser pfm gets set to STRICT in the test, then the parser is added to the context with `reset()`which changes is back to EMPTY (this is the source of the error here, will be fixed with this PR) and then it is never reset.

javanna replied:
/me tries to figure out whether this is good. I think it is, cause that means that in the context we will have the right thing depending on settings, while tests use strict mode, while before strict became empty by mistake?

cbuescher replied:
Yes, I also added a test tieh the deprecated "sort_mode" syntax to check that we throw an error in strict mode.

javanna commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java:
```diff
@@ -180,5 +182,12 @@ public ReindexParseContext(QueryParseContext queryParseContext, AggregatorParser
+			return queryParseContext.parseFieldMatcher();
+		}
+
+
```
nit: one too many line break? :)

javanna commented:
I like the change, it cleans a lot of things up, and actually unveils some more mess that will need to be cleaned up further, but that is good! left a few comments.

cbuescher committed:
Rename context.parseFieldMatcher() to context.getParseFieldMatcher

cbuescher commented:
@javanna thanks, I addressed your comments. The renaming of the getter from `parseFieldMatcher` to `getParseFieldMatcher` touches quiet a bit of code though. So I'm not sure if that should be a follow up, wdyt?

javanna commented:
LGTM, as for the rename, do what you prefer, I have no problem either way

cbuescher commented:
@javanna thanks, will merge after final test run
